### PR TITLE
Move validations before calling interpret method

### PR DESCRIPTION
**Pull request by qequ**
The method Validations that reports errors on Values was called after the Encoder's interpret function but before encValue, that's why the repeated objects error was reported when exporting to JSON but ignored when exporting to OpenAPI.  Possibly this bug affected ProtobufJSON exporting too.

Moving it before works as expected, signaling the repeated objects error before generating the OpenAPI specification.

OpenAPI tests passed.

#1715

### CODE REVIEW

**Comment by mvdan** [comment]
This fix seems reasonable, but please add a test :)

To make the DCO bot happy, see https://github.com/cue-lang/cue/blob/master/doc/contribute.md#asserting-a-developer-certificate-of-origin.

**Comment by myitcv** [comment]
Welcome to the CUE project, @qequ! 

Please see the [contribution guide](https://github.com/cue-lang/cue/blob/master/doc/contribute.md#good-commit-messages) for advice on commit messages.

**Comment by qequ** [comment]
@mvdan Added tests :+1:

**Comment by mvdan** [comment]
Ah, perhaps I should have clarified - in general, regression tests are added in the form of "testscript" files, which help test end-to-end flows without having to write separate unit tests.

For example, see `cmd/cue/cmd/testdata/script/def_openapi.txt`; you can probably create a similar and smaller sibling of it at `cmd/cue/cmd/testdata/script/export_openapi.txt`.

**Comment by qequ** [comment]
@mvdan Added testscript file

File: internal/encoding/encoder_test.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+import (
+	"testing"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/build"
+	"cuelang.org/go/cue/parser"
+)
+
+func TestEncoderValidation(t *testing.T) {
```
Comment (mpvl) [inline internal/encoding/encoder_test.go]: do we still need this test now we have the other test?

File: cmd/cue/cmd/testdata/script/export_openapi_err.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
+! cue export repeated.cue -o openapi:export.json
+cmp stdout expect-stdout
+cmp stderr expect-stderr
+
+! cue eval repeated.cue --out openapi
+cmp stdout expect-stdout
```
Comment (mvdan) [inline cmd/cue/cmd/testdata/script/export_openapi_err.txt]: you can just do `! stdout .` to verify that stdout is empty; no need for an empty expect-stdout file.

File: internal/encoding/encoder_test.go (hunk 1 above)
Comment (mvdan) [reply internal/encoding/encoder_test.go]: agreed; that's what I meant originally. this file can be deleted.

**Review by mvdan — COMMENTED** [review]
I also just realised that export_err.txt exists; maybe you could add your test case as a new sub-package there, to avoid creating a new tiny file.

**Comment by qequ** [comment]
@mvdan  @mpvl  applied changed to tests

**Comment by mvdan** [comment]
Thank you! Could you please squash your commits? Your final commit message is what we will use when importing to Gerrit to then test and merge. See https://github.com/cue-lang/cue/blob/master/doc/contribute.md#good-commit-messages, like Paul said.

**Comment by qequ** [comment]
@mvdan https://github.com/cue-lang/cue/pull/1716/commits/21576a2b2b6f3621dac80df2f33b47412bbef4a7 squashed commit

**Comment by mvdan** [comment]
I still see seven commits. Did you squash and rebase on the latest master?

I can always squash myself and use your commit message from 21576a2b2b6f3621dac80df2f33b47412bbef4a7, if you prefer.

**Comment by qequ** [comment]
@mvdan I thought I had. 

> I can always squash myself and use your commit message from https://github.com/cue-lang/cue/commit/21576a2b2b6f3621dac80df2f33b47412bbef4a7, if you prefer.

That would be great! Thank you very much 😁

**Comment by qequ** [comment]
@mvdan any updates on this issue?

**Comment by mvdan** [comment]
I still intend to merge this; it's just a slightly involved process as we have to import it to Gerrit, and it's the first time I'm doing it - while at the same time automating the process :) I'll likely wrap it up tomorrow.

**Comment by mvdan** [comment]
Attempt at https://review.gerrithub.io/c/cue-lang/cue/+/539077, if you want to give it a look. I squashed your commits and reused most of the description you gave in this PR for the commit message.

**Pull request closed without merging**